### PR TITLE
Support access log from sidecar and per-endpoint redirect stats

### PR DESCRIPTION
**Pull request by rlenglet**
Use the policy name instead of listener ID in Envoy access logs.

Remove duplicate `accesslog.pb.go` file. Move files related to access log protobufs into the right package (`pkg/envoy/cilium`). Remove `listener_id` from filter configuration. Use network policy
resource name instead.

Extract network policy name and observation point from each log record, instead of getting it from an association with a redirect / listener. Instead, associate access log producers to network policies.

Add the observation point (ingress / egress) in every Envoy log record, and use that to create an access log record.

Remove all logging logic from `Redirect`.    
Remove all stats logic from `Redirect`, to decouple stats from redirects to local proxies.

Split the `LogRecordProducer` interface into the existing `LogRecordNotifier` interface and a new `EndpointInfoRegistry` interface. Group identity lookup by ID and by IP under the same `EndpointInfoRegistry` interface.
    
Split files for readability:

- `logger.go: move endpoint info interface into `epinfo.go`
- `proxy.go`: move `Redirect` into `redirect.go`, `getMagicMark` into `mark.go`

Move redirect statistics from proxy status to endpoint API. Remove statistics from `ProxyRedirectStatus`. Add security identity and labels SHA256 to `ProxyRedirectStatus`, for consistency.
    
Move `ProxyRedirectStatistics` into `Endpoint` (i.e. `cilium endpoint get`) and redefine it as a subset of `ProxyRedirectStatus` plus statistics.

The output of `cilium endpoint get` now returns stats like:

```
        "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "statistics": {
              "requests": {
                "denied": 12,
                "forwarded": 15,
                "received": 27
              },
              "responses": {
                "forwarded": 15,
                "received": 15
              }
            }
          }
        ],
```

Rename the `NetworkPolicyEndpoint` interface into `EndpointUpdater`. Move `EndpointUpdater` from `pkg/envoy` into `pkg/proxy/logger` to make usable from the Kafka proxy.

Update stats from Kafka proxy after logging into access log.
    
Fix access logging to only log and account once when a request is denied and a response can't be created.

Add endpoint identities, IP addresses, and ports in Kafka response access logs. This is also required to be able to derive endpoint stats.

Look up original destination address and port of a Kafka request connection only once after accepting the connection, instead of for every request.

Simplify the implementation of the Kafka proxy's `handleRequests` function as it was always passed a `nil` log record.

Remove the `redirects` field from proxy status in `cilium status`. Move the `allocated-proxy-port` field into `proxy-statistics` in `cilium endpoint get`.

Sort the elements in `proxy-statistics`.

Signed-off-by: Romain Lenglet <romain@covalent.io>

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

File: pkg/proxy/logger/logger.go [hunk 1]
```diff
@@ -62,80 +62,95 @@ const (
...
+		tagFn(&lr)
 	}
+
+	return &lr
 }
 
-// fillEndpointInfo tries to resolve the IP address and fills the EndpointInfo
-// fields with either ReservedIdentityHost or ReservedIdentityWorld
-func fillEndpointInfo(producer LogRecordProducer, info *accesslog.EndpointInfo, ip net.IP) {
+// fillEndpointInfo tries to resolve the IP address and return the EndpointInfo
```
Comment (tgraf) [inline pkg/proxy/logger/logger.go]: The adjusted comment doesn't make sense

File: pkg/proxy/logger/logger.go [hunk 2]
```diff
@@ -217,11 +230,11 @@ type AddressingInfo struct {
 // to the logrecord
 func (logTags) Addressing(i AddressingInfo) LogTag {
 	return func(lr *LogRecord) {
-		switch lr.ObservationPoint {
-		case accesslog.Ingress:
-			lr.DestinationEndpoint = lr.producer.LocalEndpointInfo()
-		case accesslog.Egress:
-			lr.SourceEndpoint = lr.producer.LocalEndpointInfo()
+		ingress := lr.ObservationPoint == accesslog.Ingress
```
Comment (tgraf) [inline pkg/proxy/logger/logger.go]: Not sure this style change is really necessary

**Review by tgraf — DISMISSED** [review]
Will review the rest later

**Review by ianvernon — APPROVED** [review]
API and endpoint changes look good. Having such visibility per endpoint is an awesome addition :D

**Comment by rlenglet** [comment]
Successful build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/1574/

File: pkg/proxy/logger/logger.go (hunk 2 above)
Comment (rlenglet) [reply pkg/proxy/logger/logger.go]: It's an optimization. 1 string comparison instead of 4.
That's not essential. I reverted this.

File: pkg/proxy/logger/logger.go (hunk 1 above)
Comment (rlenglet) [reply pkg/proxy/logger/logger.go]: Reverted. This was leftover from another change I tried.

**Comment by rlenglet** [comment]
test-me-please

File: api/v1/openapi.yaml [hunk 3]
```diff
@@ -824,8 +824,13 @@ definitions:
...
         type: integer
       proxy-policy-revision:
-        description: The policy revision currently enforced in the proxy
+        description: The policy revision currently enforced in the proxy for this endpoint
         type: integer
+      proxy-statistics:
+        description: Statistics of the proxy redirects configured for this endpoint
+        type: array
+        items:
+          "$ref": "#/definitions/ProxyRedirectStatistics"
```
Comment (tgraf) [inline api/v1/openapi.yaml]: This split seems weird, why not just encode the entire proxy redirect status in the endpoint? Splitting this means a user has to manually correlate the statistics via the endpoint and rules and proxy port via the global status.

**Review by tgraf — CHANGES_REQUESTED** [review]
What is the state of Kafka with this current PR? Does this break Kafka?

File: api/v1/openapi.yaml (hunk 3 above)
Comment (rlenglet) [reply api/v1/openapi.yaml]: Because we won't have that information from sidecar proxies. The only information that we won't have is the proxy redirect port. I don't think that's essential.

Comment (rlenglet) [reply api/v1/openapi.yaml]: Kafka shouldn't break.

**Comment by rlenglet** [comment]
Kafka should still work. I will test manually to confirm.

File: api/v1/openapi.yaml (hunk 3 above)
Comment (rlenglet) [reply api/v1/openapi.yaml]: One advantage of aggregating stats at endpoint level is that we'll have stats that will span the whole lifetime of the endpoint, not just the lifetime of redirects. I think that's more useful. For instance, we can still see the total count of denied requests even after policy changes that result in redirect changes.

Comment (rlenglet) [reply api/v1/openapi.yaml]: And we can't correlate stats from sidecar proxies with redirects, because we don't even configure redirects in this case. That's the whole point of this PR.

**Comment by rlenglet** [comment]
CI test failure in Kafka test:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/1582/
```
17:37:07 K8sValidatedKafkaPolicyTest
17:37:07 /root/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:28
17:37:07   KafkaPolicies [It]
17:37:07   /root/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:115
17:37:07 
17:37:07   Expected
17:37:07       <*errors.errorString | 0xc4203ecde0>: {
17:37:07           s: "ExecKafkaPodCmd: command 'kubectl exec -n default empire-hq-55c766b8b8-hjrhx sh -- -c \"echo “Happy 40th Birthday to General Tagge” | ./kafka-produce.sh --topic empire-announce\"' failed '>>' || '[2018-03-23 17:35:43,158] WARN Got error produce response with correlation id 3 on topic-partition empire-announce-0, retrying (2 attempts left). Error: NETWORK_EXCEPTION (org.apache.kafka.clients.producer.internals.Sender)\n[2018-03-23 17:35:44,793] ERROR Error when sending message to topic empire-announce with key: null, value: 42 bytes with error: (org.apache.kafka.clients.producer.internals.ErrorLoggingCallback)\norg.apache.kafka.common.errors.TimeoutException: Expiring 1 record(s) for empire-announce-0: 1534 ms has passed since last attempt plus backoff time\n'",
17:37:07       }
17:37:07   to be nil
17:37:07 
17:37:07   /root/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:135
```

No errors in the Cilium logs. And we don't enable flow debug logs in the CI, nor collect access logs. Impossible to know what happened here.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Verified manually that Kafka access logging is working as before.
I need to fix endpoint stats for the Kafka proxy though.

**Comment by rlenglet** [comment]
Successful build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/1584/

**Comment by tgraf** [comment]
> Kafka should still work. I will test manually to confirm.

Mainly wondering what the Kafka TODO comment is about

**Comment by rlenglet** [comment]
Oh, that comment is for when we move the Kafka proxy into a separate process, and therefore use the xDS protocol for configuration and the same access log protocol. Right now we assume that only HTTP proxies are using those protocols.

**Comment by rlenglet** [comment]
I found and fixed 2 minor issues in the Kafka broker:

- We were not setting the endpoint IDs, incl. IP addresses and ports, in access logs for Kafka responses.
- We were querying the proxymap for the original destination IP and port for every Kafka request, instead of once after accepting a Kafka request connection.

I fixed both. Now we'll see response access log records with endpoint identities:
```
{"Type":"Request","Timestamp":"2018-03-23T23:11:51.14185254Z","NodeAddressInfo":{"IPv4":"10.0.2.15","IPv6":"fd00::b"},"ObservationPoint":"Ingress","SourceEndpoint":{"ID":0,"IPv4":"10.11.13.37","IPv6":"","Port":38958,"Identity":17350,"Labels":["container:id.client"],"LabelsSHA256":"c2e7b3482b5e9e1abca840b8cc5568ff876c7524d723b3068683f539008537dc"},"DestinationEndpoint":{"ID":29381,"IPv4":"10.11.193.100","IPv6":"f00d::a0f:0:0:72c5","Port":9092,"Identity":39460,"Labels":["container:id.kafka"],"LabelsSHA256":"e0c546c60fd39233aaf59dc8d8f63a43145713aa8bd24a172f15ff6162c9d045"},"IPVersion":0,"Verdict":"Forwarded","Info":"","Metadata":null,"TransportProtocol":6,"FlowEvent":"","ServiceInfo":null,"DropReason":null,"Kafka":{"ErrorCode":0,"APIVersion":3,"APIKey":"offsetcommit","CorrelationID":62,"Topic":{"Topic":"allowedTopic"}}}
{"Type":"Response","Timestamp":"2018-03-23T23:11:51.147209563Z","NodeAddressInfo":{"IPv4":"10.0.2.15","IPv6":"fd00::b"},"ObservationPoint":"Ingress","SourceEndpoint":{"ID":0,"IPv4":"10.11.13.37","IPv6":"","Port":38958,"Identity":17350,"Labels":["container:id.client"],"LabelsSHA256":"c2e7b3482b5e9e1abca840b8cc5568ff876c7524d723b3068683f539008537dc"},"DestinationEndpoint":{"ID":29381,"IPv4":"10.11.193.100","IPv6":"f00d::a0f:0:0:72c5","Port":9092,"Identity":39460,"Labels":["container:id.kafka"],"LabelsSHA256":"e0c546c60fd39233aaf59dc8d8f63a43145713aa8bd24a172f15ff6162c9d045"},"IPVersion":0,"Verdict":"Forwarded","Info":"","Metadata":null,"TransportProtocol":6,"FlowEvent":"","ServiceInfo":null,"DropReason":null,"Kafka":{"ErrorCode":0,"APIVersion":3,"APIKey":"offsetcommit","CorrelationID":62,"Topic":{"Topic":"allowedTopic"}}}
```

**Comment by rlenglet** [comment]
And now we'll see Kafka stats in `cilium endpoint get`:
```
    "proxy-statistics": [
      {
        "location": "ingress",
        "port": 9092,
        "protocol": "kafka",
        "statistics": {
          "requests": {
            "forwarded": 80,
            "received": 80
          },
          "responses": {
            "forwarded": 79,
            "received": 79
          }
        }
      }
    ],
```

**Comment by rlenglet** [comment]
test-me-please

**Review by jrajahalme — APPROVED** [review]
LGTM

File: pkg/proxy/kafka.go [hunk 4]
```diff
@@ -322,20 +346,17 @@ func handleRequests(done <-chan struct{}, pair *connectionPair, c *proxyConnecti
 		}
 
 		if err != nil {
-			if record != nil {
```
Comment (manalibhutiyani) [inline pkg/proxy/kafka.go:325]: Why was this removed? Shouldn't we be putting this in access log?

Comment (rlenglet) [reply pkg/proxy/kafka.go:325]: record was always nil. This was dead code.

File: api/v1/openapi.yaml (hunk 3 above)
Comment (tgraf) [reply api/v1/openapi.yaml]: I agree that having the stats in the endpoint is neat. I also understand that there is no conventional proxy port, although it would be great to see the port number on which Envoy is listening. My main concern left is that in order to see the proxy rules, the user has to correlate the entry in the status.

Users will also get used to seeing the proxy status in either place and it will just be empty for the sidecar model. A user will not want to understand what a redirect and why it is missing for the sidecar case. Can't we move everything to the endpoint and just not call it a redirect?

**Comment by rlenglet** [comment]
Successful build: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/1597/

**Comment by rlenglet** [comment]
test-me-please

File: api/v1/openapi.yaml (hunk 3 above)
Comment (rlenglet) [reply api/v1/openapi.yaml]: We can't get the redirect port from sidecars we don't manage, e.g. Istio sidecars. And even if we did it wouldn't be useful information. That port is 15000 by default for every redirect across the whole cluster.

I could add the redirect port into the Endpoint's `proxy-statistics` if there's a redirect currently active for the port *and* if the redirect port is known:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "redirect-port": 12345,
            "statistics": {
              ...
            }
          }
        ],
```

And if there's currently no redirect for the port, or the redirect port is not known, there will be no `"redirect-port"` field:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "statistics": {
              ...
            }
          }
        ],
```

And then I can remove the redirects from `cilium status` altogether.

The only drawback is that it will be hard to know at a glance what redirects are currently installed into the Envoy and Kafka proxies managed by Cilium (not in sidecars).

@tgraf, would you be OK with that API?

Comment (rlenglet) [reply api/v1/openapi.yaml]: And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by `cilium endpoint get`.

When we move the Kafka proxy to use xDS / NPDS for configuration, that will also be true for Kafka. We will NOT configure rules specifically for each redirect.

**Comment by rlenglet** [comment]
test-me-please

File: api/v1/openapi.yaml (hunk 3 above)
Comment (tgraf) [reply api/v1/openapi.yaml]: > That port is 15000 by default for every redirect across the whole cluster.

It may be clear to you, it will not be clear to users and it will be useful information to the user as there will be traffic inside of the pod to and from that port. As long as we can add it later on, I'm fine.

> And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by cilium endpoint get.

That's great, given we are just before 1.0, we absolutely don't want to expose something in the overall status and then remove it. How difficult is it to just expose the kafka rules like this as well? I really dislike exposing any information in different places depending on whether the proxy model is sidecar or not.

Comment (rlenglet) [reply api/v1/openapi.yaml]: We'll have to move the Kafka proxy to use NPDS when we move it into a separate process. When that's done, the policy that is returned in `cilium endpoint get` will be the one that is pushed to the Kafka proxy.

I'll remove `redirects` altogether from the proxy status then.

**Comment by rlenglet** [comment]
Test failure in this build:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/1611/
```
20:00:05     not to contain substring
20:00:05         <string>: POTENTIAL DEADLOCK:
20:00:05 
20:00:05     /root/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/runtime/connectivity.go:41
```

It seems like there is a panic in the xDS server:
```
Mar 24 19:59:15 runtime cilium-agent[28686]: POTENTIAL DEADLOCK: Duplicate locking, saw callers this locks in one goroutine:
Mar 24 19:59:15 runtime cilium-agent[28686]: current goroutine 433 lock &{{{1 0} 0 0 -1073741824 0}}
Mar 24 19:59:15 runtime cilium-agent[28686]: all callers to this lock in the goroutine
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:520 endpoint.(*Endpoint).regenerateBPF { e.Mutex.Lock() } <<<<<
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:844 endpoint.(*Endpoint).regenerate { revision, err := e.regenerateBPF(owner, tmpDir, reason) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:918 endpoint.(*Endpoint).Regenerate.func1 { if err := e.regenerate(owner, reason); err != nil { }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:839 endpoint.(*Endpoint).regenerate.func2 { e.Mutex.Lock() } <<<<<
Mar 24 19:59:15 runtime cilium-agent[28686]: /usr/local/go/src/runtime/asm_amd64.s:509 runtime.call32 { CALLFN(·call32, 32) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /usr/local/go/src/runtime/panic.go:491 runtime.gopanic { reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz)) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /usr/local/go/src/runtime/panic.go:63 runtime.panicmem { panic(memoryError) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /usr/local/go/src/runtime/signal_unix.go:367 runtime.sigpanic { panicmem() }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/envoy/server.go:459 envoy.getNetworkPolicy { func getNetworkPolicy(name string, id identity.NumericIdentity, policy *policy.L4Policy, }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/envoy/server.go:496 envoy.(*XDSServer).UpdateNetworkPolicy { networkPolicy := getNetworkPolicy(ip, ep.GetIdentity(), policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/daemon/daemon.go:156 main.(*Daemon).UpdateNetworkPolicy { return d.l7Proxy.UpdateNetworkPolicy(e, policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities, e.ProxyWaitGroup) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:589 endpoint.(*Endpoint).updateNetworkPolicy { err := owner.UpdateNetworkPolicy(e, c.L4Policy, *e.LabelsMap, deniedIngressIdentities, nil) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:666 endpoint.(*Endpoint).regenerateBPF { if err = e.updateNetworkPolicy(owner); err != nil { }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:844 endpoint.(*Endpoint).regenerate { revision, err := e.regenerateBPF(owner, tmpDir, reason) }
Mar 24 19:59:15 runtime cilium-agent[28686]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:918 endpoint.(*Endpoint).Regenerate.func1 { if err := e.regenerate(owner, reason); err != nil { }
```

Will look into that separately. It's not related to this PR.

This panic seems caused by an invalid dereference. The only dereferences done in this function (`envoy.getNetworkPolicy`) is on the endpoint's `policy`. That's easy to fix. If `policy` is `nil`, return a deny-all policy. I'll fix that in a small separate PR.

**Comment by rlenglet** [comment]
Updated the endpoint API to include the proxy redirect port when known, and sorted the stats by port, protocol, etc.:
```
    "proxy-statistics": [
      {
        "allocated-proxy-port": 17725,
        "location": "ingress",
        "port": 80,
        "protocol": "http",
        "statistics": {
          "requests": {
            "denied": 6,
            "forwarded": 9,
            "received": 15
          },
          "responses": {
            "forwarded": 9,
            "received": 9
          }
        }
      },
      {
        "allocated-proxy-port": 16699,
        "location": "ingress",
        "port": 8000,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      },
      {
        "allocated-proxy-port": 15133,
        "location": "ingress",
        "port": 8080,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      }
    ],
```

When the redirect port is not known or the redirect is removed, only the `allocated-proxy-port` is removed:
```
    "proxy-statistics": [
      {
        "location": "ingress",
        "port": 80,
        "protocol": "http",
        "statistics": {
          "requests": {
            "denied": 11,
            "forwarded": 16,
            "received": 27
          },
          "responses": {
            "forwarded": 16,
            "received": 16
          }
        }
      },
      {
        "location": "ingress",
        "port": 8000,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      },
      {
        "location": "ingress",
        "port": 8080,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      }
    ],
```

**Comment by rlenglet** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
